graph/api/src/main/scala/org/apache/spark/graph/api/CypherSession.scala

```diff
@@ -0,0 +1,285 @@
+
+  /**
+   * Executes a Cypher query on the given input graph.
+   *
+   * Note that queries can take optional parameters:
+   *
+   * {{{
+   *     Parameters:
+   *
+   *     {
+   *        "name" : "Alice"
+   *     }
+   *
+   *     Query:
+   *
+   *     MATCH (n:Person)
+   *     WHERE n.name = $name
+   *     RETURN n
+   * }}}
+   *
+   * @param graph      [[PropertyGraph]] on which the query is executed
+   * @param query      Cypher query to execute
+   * @param parameters parameters used by the Cypher query
+   * @since 3.0.0
+   */
+  def cypher(
+      graph: PropertyGraph,
+      query: String,
+      parameters: java.util.Map[String, Object]): CypherResult = {
+    cypher(graph, query, parameters.asScala.toMap)
+  }
+
+  /**
+   * Creates a [[PropertyGraph]] from a sequence of [[NodeDataset]]s and [[RelationshipDataset]]s.
+   * At least one [[NodeDataset]] has to be provided.
+   *
+   * For each label set and relationship type there can be at most one [[NodeDataset]] and at most
+   * one [[RelationshipDataset]], respectively.
+   *
+   * @param nodes         NodeDataset that define the nodes in the graph
+   * @param relationships RelationshipDataset that define the relationships in the graph
+   * @since 3.0.0
+   */
+  def createGraph(
+      nodes: Array[NodeDataset],
+      relationships: Array[RelationshipDataset]): PropertyGraph
+
+  /**
+   * Creates a [[PropertyGraph]] from nodes and relationships.
+   *
+   * The given dataset needs to adhere to the following column naming conventions:
+   *
+   * {{{
+   *     Id column:        `$ID`            (nodes and relationships)
+   *     SourceId column:  `$SOURCE_ID`     (relationships)
+   *     TargetId column:  `$TARGET_ID`     (relationships)
+   *
+   *     Label columns:    `:{LABEL_NAME}`  (nodes)
+   *     RelType columns:  `:{REL_TYPE}`    (relationships)
+   *
+   *     Property columns: `{Property_Key}` (nodes and relationships)
+   * }}}
+   *
+   * @note It is recommended to cache the input datasets if they represent multiple label sets and
+   *       relationship types.
+   *
+   * @see [[CypherSession]]
+   * @param nodes         node dataset
+   * @param relationships relationship dataset
+   * @since 3.0.0
+   */
+  def createGraph(nodes: Dataset[Row], relationships: Dataset[Row]): PropertyGraph = {
+    val nodeFrames = CypherSession.extractNodeDatasets(nodes)
+    val relationshipFrames = CypherSession.extractRelationshipDatasets(relationships)
+    createGraph(nodeFrames.toArray, relationshipFrames.toArray)
+  }
+
+  /**
+   * Returns a [[PropertyGraphReader]] that can be used to read data in as a `PropertyGraph`.
+   *
+   * @since 3.0.0
+   */
+  def read: PropertyGraphReader
+
+  /**
+   * Returns a [[NodeDatasetBuilder]] that can be used to construct a [[NodeDataset]].
+   *
+   * @param ds Dataset containing a single node in each row
+   * @since 3.0.0
+   */
+  def buildNodeDataset(ds: Dataset[Row]): NodeDatasetBuilder =
```

Comment:
It seems the construction of `NodeDataset` has no dependency on `CypherSession`. So we shouldn't define it here. Any corner cases I missed?

Reply:
Are you saying we should use the builder to create the `Node/RelDatasets` or are you saying we should move the convenience methods to some place else?

Reply:
* User can constructor the builder directly. Say if I want to construct `NodeDataset`, I will find constructors or builders in `NodeDataset` API doc, instead of `CypherSession`.
* It is also a little confusing that `buildNodeDataset` doesn't build it but returns a builder. `NodeDataset.builder` is much clearer.